File: Frontend/src/services/commentService.ts
// commentService.ts
import api from './api';

export interface Comment {
  id: string;
  taskId: string;
  authorId: string;
  content: string;
  parentId?: string;
  createdAt: string;
  updatedAt: string;
}

class CommentService {
  private basePath = '/api/v1/taskfarm/comments';

  public async getByTask(taskId: string): Promise<Comment[]> {
    return await api.get<Comment[]>(`${this.basePath}/task/${taskId}`);
  }

  public async getById(id: string): Promise<Comment> {
    return await api.get<Comment>(`${this.basePath}/${id}`);
  }

  public async create(comment: Partial<Comment>): Promise<Comment> {
    return await api.post<Comment>(this.basePath, comment);
  }

  public async update(id: string, comment: Partial<Comment>): Promise<Comment> {
    return await api.put<Comment>(`${this.basePath}/${id}`, comment);
  }

  public async delete(id: string): Promise<void> {
    return await api.delete(`${this.basePath}/${id}`);
  }

  public async reply(parentId: string, taskId: string, authorId: string, content: string): Promise<Comment> {
    return await this.create({ taskId, authorId, content, parentId });
  }
}

const commentService = new CommentService();
export default commentService;

File: Backend/src/main/java/ir/netpick/platform/taskfarm/service/CommentService.java
package ir.netpick.platform.taskfarm.service;

import ir.netpick.platform.core.PageDTO;
import ir.netpick.platform.core.constants.GeneralConstants;
import ir.netpick.platform.core.exception.ResourceNotFoundException;
import ir.netpick.platform.core.utils.PageDTOMapper;
import ir.netpick.platform.taskfarm.model.Comment;
import ir.netpick.platform.taskfarm.repository.CommentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CommentService {

    private final CommentRepository commentRepository;

    public List<Comment> getByTask(UUID taskId) {
        return commentRepository.findByTaskIdAndDeletedFalseOrderByCreatedAtAsc(taskId);
    }

    public PageDTO<Comment> getAll(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<Comment> page = commentRepository.findByDeletedFalse(pageable);
        return PageDTOMapper.map(page);
    }

    public Comment getById(UUID commentId) {
        return commentRepository.findById(commentId)
                .orElseThrow(() -> new ResourceNotFoundException("Comment with id [%s] not found".formatted(commentId)));
    }

    public Comment create(Comment comment) {
        comment.setId(null);
        return commentRepository.save(comment);
    }

    public Comment update(UUID commentId, Comment comment) {
        Comment existing = getById(commentId);
        comment.setId(commentId);
        comment.setCreatedAt(existing.getCreatedAt());
        return commentRepository.save(comment);
    }

    public void delete(UUID commentId) {
        commentRepository.deleteById(commentId);
    }
}